### ISSUE TITLE
PeriodicTorsionForce does not support wildcard when Types are used instead of Classes 

### ISSUE THREAD

**Issue by rafwiewiora** [pandegroup/openmm#1347]
Spin out of a discovery in https://github.com/ParmEd/ParmEd/issues/486 that ffxml written using Types instead of Classes for PeriodicTorsionForce gives wrong energies. 

Tracked this down to the lack of support for wildcard when types are used. 

Only `Forcefield._atomClasses` contains the `''` key: https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L233

Type attribute using `''` will return a `None` type: https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L292

and consequently the torsion parser:
https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L333

will cause the `PeriodicTorsionGenerator` not to register those torsions. 

PR coming soon!

**Comment by jchodera** [comment]
This may be a good opportunity to generalize how we use attributes like `type` and `class` in parameter matching.  I think the current scheme---where [`type` is appended to a list for a given `class` as the type is registered](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L227-L232)---is highly limiting, in addition to creating bugs (#1344).

In addition, some of the force generators match explicitly by type or class only, or handle them differently.  All of the Amoeba forces assume that only `class` is used, and only use `class` attributes in their error messages (e.g. [here](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L4441-L4443) and [here](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L3608-L3614)).

What if we instead refactored `ForceField` a bit to isolate the atom type checking logic, allowing us to make it more extensible for the future?  This might allow us to simplify even simple cases like [`HarmonicBondForce`](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/forcefield.py#L1037-L1049) as well as make everything behave more uniformly.

**Comment by jchodera** [comment]
We can of course fix the lack of wildcards for now if that's easy to do.

**Comment by rafwiewiora** [comment]
Yes, I agree this is a good moment to make some bigger changes here. Leaving this open to discussion. 

However, also pushing a quick fix for the type wildcard problem: https://github.com/pandegroup/openmm/pull/1348 as this probably shouldn't wait - if I understand the docs correctly, types should be fully exchangeable with classes, so this is a bad bug we should fix for 7.0.

**Comment by rafwiewiora** [comment]
Ok, so changed the topic here to more general discussion about class and type. @jchodera what exactly are your proposing going forward?

**Comment by jchodera** [comment]
I'll have a more concrete proposal in a day or so after I dig into the `ForceGenerator`s a bit more.

**Comment by rafwiewiora** [comment]
Anything new here?

**Comment by jchodera** [comment]
Sorry about the delay.  I think we need to make this sufficiently general so that `type` and `class` both act in similar manners, and both allow wildcards to be specified.

I don't have a strong feeling about whether parameter matching should be allowed to mix-and-match `type` and `class` in the same parameter definition, but I *do* feel that all of the `ForceGenerators` should support the *same rules* for matching by `type` and `class`.  Currently, this is highly inconsistent between `ForceGenerators`, with some of them (mostly AMOEBA ones) that only support `class`.

Any revisions here should likely go together with revisions that allow all files to be loaded before `class` is expanded, though I think the code for matching atoms can simply be refactored out into a central location so that behavior can be ensured to be identical across `ForceGenerators`.

**Comment by peastman** [comment]
All generators should support both type and class.  If any of them don't, that's a bug.

**Comment by peastman** [comment]
Was this ever resolved?  Did we determine whether there's an actual problem, and if so, what it is?
